Makefile

```diff
@@ -1898,6 +1898,9 @@ clipping_iterator_test: $(OBJ_DIR)/db/compaction/clipping_iterator_test.o $(TEST
 ribbon_bench: $(OBJ_DIR)/microbench/ribbon_bench.o $(LIBRARY)
 	$(AM_LINK)
 
+cache_rep_test: $(OBJ_DIR)/cache/cache_rep_test.o $(TEST_LIBRARY) $(LIBRARY)
```

Comment:
Interesting! I never thought of that aspect at all! Does "new test binary" refer to the "cache_rep_test" executable produced after the command "make"?

My sole reason of adding the line 1901 here is to create a rule in Makefile so that I can build and run this specific test by doing "make -jN cache_rep_test" and "./cache_rep_test". This also allow others to do the same thing. The ultimate goal here is to facilitate a smaller hence faster testing, for people just interested in testing my cache_rep.

However, there could be other ways to achieve a similar purpose as above that don't need to "add new test binaries" that I don't know of. But I am more than happy to learn about them :)

Reply:
Interesting! This aspect didn't even occur to my mind when I was adding the line, indicating that there is something more in the RocksDB's building process I need to learn :)

So here is a learning question: I was under the impression that the new test binaries (in my case, cache_rep_test executable) are created only under after running: 
(1)"make check" or "make all". The executable will be created regardless of whether line 1901 exists since "cache_rep_test.cc" is in "TEST_MAIN_SOURCE" in "src.mk". 
Or (2)"make cache_rep_test". The rule "cache_rep_test" is valid and the executable will be created after running the rule because I added line 1901. 

Both (1), (2) are intentional commands that are run by people who probably want to run the "cache_rep_test" executable for testing. If people don't run (1) nor (2), no "cache_rep_test test" executable (binary) will be created/built. (This is the part I am not sure about - is there any other automatic mechanism going on that will automatically create new "cache_rep_test" executable/binary even when people don't want to run "make check/make all/make cache_rep_test", just because I added "cache_rep_test.cc" is in "TEST_MAIN_SOURCE" in "src.mk" and added line 1901?)

Also just for me to make sure I understand your comment - are you suggesting to (1)take out "cache_rep_test.cc" from src.mk OR/AND (2)take out the line 1901 in the MakeFile?

Thanks!

Reply:
I'm suggesting adding new tests to an existing test binary and existing test source file unless you have a compelling reason to create a new one. Such a reason would be a necessarily large amount of new testing.

Reply:
If each header file had its own test binary, we would have 500+ test binaries, which would build 100+GB of test binaries. We could potentially switch to LIB_MODE=shared by default, but that's only a partial mitigation.

Reply:
[non-blocking] [NEED_ATTENTION]
@ajkr - I agree with what @pdillinger mentioned above and tried to look for an existing test file, where I can place my unit test for the new class CacheRep but not awkwardly.

cache_test.cc and lru_cache_test.cc don't seem to be good ones, as the unit tests in those files have coherent themes (i.e, testing the cache or testing the lru_cache). It seems awkward to put my cache_rep_test into any of those files.

Any other test files I should look into?

Reply:
> I thought static linking would drop unreferenced symbols

Nope. Need `strip` for that. 

> Any other test files I should look into?

Maybe db_block_cache_test

Reply:
I don't know, it's 18MB for me. Maybe it's too easy to accidentally pull in everything because even simple tests like event_logger_test are 200MB+.

```
$ git apply 8506.diff && make -j48 cache_rep_test && ls -l cache_rep_test
...
-rwxr-xr-x 1 andrewkr users 18320624 Jul 15 15:43 cache_rep_test
```

Anyways move to separate file sounds fine. But tests named "db_*" are supposed to be higher-level tests using the `DB` API. How about "cache/cache_test.cc"?